kore/src/Kore/Internal/TermLike/TermLike.hs

```diff
@@ -1249,3 +1262,136 @@ depth :: TermLike variable -> Int
+            wrap $
+                PatternF.ApplicationF $
+                    first Symbol.toSymbolOrAlias $ Signedness.toApplication signedness
+        InjF inj ->
+            wrap $
+                PatternF.ApplicationF $
+                    first Symbol.toSymbolOrAlias $ Inj.toApplication inj
+      where
+        wrap x = CofreeT $ Identity $ attr :< x
+        mkInternalDV sort strVal =
+            PatternF.DomainValueF $
+                DomainValue sort $
+                    wrap $
+                        PatternF.StringLiteralF $ Const $ StringLiteral strVal
+
+        mkTrm unitSymbol concatSymbol = mkAux
+          where
+            mkAux = \case
+                [] ->
+                    wrap $
+                        PatternF.ApplicationF $
+                            Application (Symbol.toSymbolOrAlias unitSymbol) []
+                [x] -> x
+                (x : xs) ->
+                    wrap $
+                        PatternF.ApplicationF $
+                            Application (Symbol.toSymbolOrAlias concatSymbol) [x, mkAux xs]
+
+        mkInternalList InternalList{internalListUnit, internalListConcat, internalListElement, internalListChild} =
+            mkTrm internalListUnit internalListConcat mkChildren
+          where
+            mkChildren =
+                map
+                    ( \e ->
+                        wrap $
+                            PatternF.ApplicationF $
+                                Application (Symbol.toSymbolOrAlias internalListElement) [e]
+                    )
+                    $ toList internalListChild
+
+        mkInternalAc ::
+            forall normalized.
+            AcWrapper normalized =>
+            InternalAc Key normalized (Cofree (PatternF.PatternF variable) (TermAttributes variable)) ->
+            Cofree (PatternF.PatternF variable) (TermAttributes variable)
+        mkInternalAc InternalAc{builtinAcUnit, builtinAcConcat, builtinAcChild, builtinAcElement} =
+            mkTrm builtinAcUnit builtinAcConcat mkChildren
+          where
+            NormalizedAc{elementsWithVariables, concreteElements, opaque} = unwrapAc builtinAcChild
+
+            mkChildren =
+                map
+                    ( wrap
+                        . PatternF.ApplicationF
+                        . Application (Symbol.toSymbolOrAlias builtinAcElement)
+                        . elementToApplicationArgs
+                    )
+                    elementsWithVariables
+                    ++ map
+                        ( \(k, v) ->
+                            wrap $
+                                PatternF.ApplicationF $
+                                    Application (Symbol.toSymbolOrAlias builtinAcElement) $
+                                        uninternalizeKey k :
+                                        concreteElementToApplicationArgs v
+                        )
+                        (HashMap.toList concreteElements)
+                    ++ opaque
```

Comment:
For some reason my review for this part of the code vanished.

This is very hard to read, I'd keep the combining of the uninternalized parts at the top, while the transformations themselves should be moved to a lower level since they're just implementation details.

Also, we need to document what output this produces. I mean, AC collections are, well, associative and commutative therefore they can have multiple, equivalent, uninternalized forms. We need to document that this produces uninternalized patterns of the <insert-here> form.

I also want to see unit tests for this and the list one.

Reply:
Ofc, feel free to move the AC uninternalizer and the list uninternalizer to the global scope in order to be able to test them directly.